packages/flutter/test/widgets/editable_text_test.dart

```diff
@@ -15444,6 +15444,64 @@ testWidgets('Floating cursor ending with selection', (WidgetTester tester) async
       );
     });
 
+    testWidgets('Selection is updated when the field has focus and the new selection is invalid', (WidgetTester tester) async {
+      // Regression test for https://github.com/flutter/flutter/issues/120631.
+      final TextEditingController controller = TextEditingController();
+      controller.text = 'Text';
+      final FocusNode focusNode = FocusNode();
+
+      await tester.pumpWidget(
+        MaterialApp(
+          home: EditableText(
+            key: ValueKey<String>(controller.text),
+            controller: controller,
+            focusNode: focusNode,
+            style: Typography.material2018().black.titleMedium!,
+            cursorColor: Colors.blue,
+            backgroundCursorColor: Colors.grey,
+          ),
+        ),
+      );
+
+      expect(focusNode.hasFocus, isFalse);
+      expect(
+        controller.selection,
+        const TextSelection.collapsed(offset: -1),
+      );
+
+      // Tab to focus the field.
+      await tester.sendKeyEvent(LogicalKeyboardKey.tab);
+      await tester.pumpAndSettle();
+      expect(focusNode.hasFocus, isTrue);
+      expect(
+        controller.selection,
+        kIsWeb
+          ? TextSelection(
+              baseOffset: 0,
+              extentOffset: controller.text.length,
+            )
+          : TextSelection.collapsed(
+              offset: controller.text.length,
+            ),
+      );
+
+      // Update text with an invalid selection.
```

Comment:
This comment isn't exactly what you're doing, is it old?

Reply:
I can make this comment more self explanatory, but it was done on purpose because `TextEditingController.test` setter explicitly set the selection offset to -1 (which is the root of the issue addressed in this PR): https://github.com/flutter/flutter/blob/f98499f9bbd186b2007c80898b57b5626be84785/packages/flutter/lib/src/widgets/editable_text.dart#L224-L230

Because `TextEditingController.test` is a public API, I did not focus on trying to change its implementation as it will  probably be a breaking change and also because there might be no simple fix that would work in all cases. What can be done at the `TextEditingController.test` level is probably to enhance its documentation to discourage its use and to recommand using `TextEditingController.value` instead.

While writing this, I wonder if we can add a method to `TextEditingController`, with a String parameter, which will set the text using the given value AND set the selection collapsed at the end of the text?

Reply:
Thanks 👍 .  I've shied away from adding new methods to TextEditingController because it's used so widely.  I think we do need a better place to put typical text editing operations, though.